### PR TITLE
[dynamo] Activation checkpointing as higher order op

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101028

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/torch.py [hunk 1]
```diff
@@ -813,7 +813,10 @@ def call_function(
         )
         from .builder import wrap_fx_proxy
 
-        assert kwargs is None or len(kwargs) == 0, "kwargs are not supported, yet"
+        if kwargs and len(kwargs):
+            for value in kwargs.values():
+                if not isinstance(value, ConstantVariable):
+                    assert "not constant kwargs are not supported yet"
```
Comment (voznesenskym) [inline torch/_dynamo/variables/torch.py:819]:
```suggestion
       assert all(isinstance(value, ConstantVariable) for value in kwargs.values()) or not kwargs
```

File: torch/_dynamo/variables/torch.py [hunk 2]
```diff
@@ -1242,14 +1245,15 @@ def speculate_branch(branch):
         else:
             unimplemented(f"HigherOrderOperator {self.value.__name__}")
 
+        p_kwargs = {k: v.as_proxy() for k, v in kwargs.items()}
```
Comment (voznesenskym) [inline torch/_dynamo/variables/torch.py:1248]: This is a less safe proxy_args_kwargs. As you have p_args done, just do p_kwargs/

```suggestion
        p_kwargs = proxy_args_kwargs([], kwargs) 
```

File: torch/_higher_order_ops/wrap.py [hunk 3]
```diff
@@ -9,5 +10,23 @@ def __call__(self, func, *args):
...
+    """
+    def __init__(self):
+        super().__init__("wrap_activation_checkpoint")
+
+    def __call__(self, function, *args, **kwargs):
+        # use_reentrant is set to False because this op is going to be traced.
+        # And we ensure that AOT Autograd traces through the non reentrant
+        # version of checkpointing.
+        kwargs["use_reentrant"] = False
+        result = checkpoint(function, *args, **kwargs)
```
Comment (voznesenskym) [inline torch/_higher_order_ops/wrap.py:29]: Can you add a test where this line throws? I am wondering if we handle this correctly.

File: torch/_dynamo/utils.py [hunk 4]
```diff
@@ -1589,3 +1593,37 @@ def convert(obj):
...
+        # TODO - functionalize_rng_ops flags cannot be turned ON by default
+        # because 1) Performance concerns - seed and offset are read and passed
+        # to each AOT graph 2) Inductor has rand-specific optimizations and
+        # there is work remaining to compose them together with
+        # functionalization.
+        #
+        # Until we make it ON by default, we will have to ask users to turn on
+        # this flag manually.  TODO - Revisit if there is a simpler way to
+        # resolve this problem.
+        if not torch._functorch.config.functionalize_rng_ops:
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:1623]: nit: 1 line w/ and?

File: torch/_dynamo/eval_frame.py [hunk 5]
```diff
@@ -1248,6 +1249,18 @@ def patch():
...
+        # looks likes this
+        #  1) TorchDynamo tries to wrap utils.checkpoint in a HigherOrderOp by
+        #     speculatively checking if the forward function is safe to trace.
+        #  2) If yes, then Dynamo-generated Fx graph has the wrapped higher
+        #     order op. As a result, TorchDynamo does not look inside utils.checkpoint.
+        #  3) If not, then TorchDynamo falls back to eager by performing a graph
+        #     break. And here, the following disable wrapper ensures that
+        #     TorchDynamo does not trigger again on the frames created by
+        #     utils.checkpoint innards.
+        torch.utils.checkpoint.checkpoint = disable(torch.utils.checkpoint.checkpoint)
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:1262]: I knew this was TorchPatcher before I clicked expand on GH ;) Thanks for the nice comment.

**Review by voznesenskym — APPROVED** [review]
Overall, logic LGTM. A few nits, and a request for a test where checkpointing raises in `__call__`

File: test/dynamo/test_higher_order_ops.py [hunk 6]
```diff
@@ -417,6 +430,135 @@ def f(x):
...
+            count_ops, freq=1, op=torch.ops.rngprims.philox_rand.default
+        )
+        backend = aot_autograd(fw_compiler=fw_compiler, bw_compiler=bw_compiler)
+        self._validate(
+            fn, backend, x, y, skip_check=True
+        )  # dropout decomp is known to diverge with eager
+
+    @requires_cuda()
+    @torch._functorch.config.patch(functionalize_rng_ops=True)
+    def test_fallback(self):
```
Comment (anijain2305) [inline test/dynamo/test_higher_order_ops.py:509]: @voznesenskym DId you mean this test? Here, I the checkpoint function has a graph break, and the test ensures that TorchDynamo Fx graph does not have any ops from the inside of the checkpoint function.

Comment (voznesenskym) [reply test/dynamo/test_higher_order_ops.py:509]: Stricter :) Not fallback w/ graph break, but actually raising from checkpoint.

File: test/dynamo/test_higher_order_ops.py [hunk 7]
```diff
@@ -417,6 +430,135 @@ def f(x):
...
+class ActivationCheckpointingTests(torch._dynamo.test_case.TestCase):
+    def _validate(self, fn, backend, *args, skip_check=False, fullgraph=True):
+        expected = fn(*args)
+        expected.sum().backward()
+
+        result = torch.compile(fn, fullgraph=fullgraph, backend=backend)(*args)
+        result.sum().backward()
+
+        if not skip_check:
+            self.assertEqual(result, expected)
```
Comment (zou3519) [inline test/dynamo/test_higher_order_ops.py:442]: Do we want to check that the gradients are the same? Maybe by using torch.autograd.grad instead of .backward()?

File: torch/_dynamo/utils.py (hunk 4 above)
Comment (zou3519) [inline torch/_dynamo/utils.py:1623]: nit: might be better if the config flag is named `compile_checkpoint` or something, it is not obvious at first glance why a feature flag to enable/disable torch.compile on activation checkpointing is called functionalize_rng_ops

File: torch/_dynamo/variables/torch.py [hunk 8]
```diff
@@ -1211,7 +1214,7 @@ def speculate_branch(branch):
             example_value = deepcopy_to_fake_tensor(example_res, tx.fake_mode)
 
             p_args = (lowered_node,) + p_args
-        elif self.value.__name__ == "wrap":
+        elif self.value.__name__ in ("wrap", "wrap_activation_checkpoint"):
```
Comment (zou3519) [inline torch/_dynamo/variables/torch.py:1217]: (no action required) There should be a check for side effects somewhere (that is, if the function passed to checkpoint or wrap mutates global state, then we have a problem). There isn't yet because we haven't added it because it is kind of complicated. Since wrap_activation_checkpoint is hidden behind a feature flag, I'm OK with this happening later (it's somewhere on the list of followups for wrap, https://github.com/pytorch/pytorch/issues/100278)

File: torch/_higher_order_ops/wrap.py [hunk 9]
```diff
@@ -9,5 +10,23 @@ def __call__(self, func, *args):
...
+    checkpointed (the first arg to the utils.checkpoint() function).
+    """
+    def __init__(self):
+        super().__init__("wrap_activation_checkpoint")
+
+    def __call__(self, function, *args, **kwargs):
+        # use_reentrant is set to False because this op is going to be traced.
+        # And we ensure that AOT Autograd traces through the non reentrant
+        # version of checkpointing.
+        kwargs["use_reentrant"] = False
```
Comment (zou3519) [inline torch/_higher_order_ops/wrap.py:28]: One other way of doing it is that if use_reentrant=True, then we graph break. IDK what's better here.

File: torch/_dynamo/utils.py [hunk 10]
```diff
@@ -1589,3 +1593,37 @@ def convert(obj):
...
+            raise RuntimeError(
+                "torch.compile on activation checkpointing is an experimental feature. "
+                "Please manually set torch._functorch.config.functionalize_rng_ops=True "
+                "to run torch.compile with activation checkpointing."
+            )
```
Comment (zou3519) [inline torch/_dynamo/utils.py:1628]: should this be an error or should it be raise Unsupported?

I'm not sure what currently happens with torch.utils.checkpoint.checkpoint under dynamo, but if it is that it succeeds by falling back to eager-mode PyTorch, then changing it to be a RuntimeError is probably not the best UX

File: torch/_higher_order_ops/wrap.py (hunk 3 above)
Comment (zou3519) [reply torch/_higher_order_ops/wrap.py:29]: +1, curious about this

**Review by zou3519 — APPROVED** [review]
This is cool! I don't have too much to add, just some minor nits and questions

File: torch/_higher_order_ops/wrap.py (hunk 9 above)
Comment (anijain2305) [reply torch/_higher_order_ops/wrap.py:28]: Will keep use_reentrant=False. Idea is that we will mostly use checkpoint higher order op for tracing. And in that case, it does not matter to the user if we are tracing entrant vs non-reentrant. We will support non-reentrant version with AOT Autograd tracing.

File: torch/_higher_order_ops/wrap.py (hunk 3 above)
Comment (anijain2305) [reply torch/_higher_order_ops/wrap.py:29]: Hmm, not sure what we are checking here?

~~~
        def gn(x, y):
            raise RuntimeError("Injecting error")
            return torch.sigmoid(torch.matmul(x, y))

        def fn(x, y):
            return torch.cos(torch.utils.checkpoint.checkpoint(gn, torch.sin(x), y))

        x = torch.randn(4, 4, requires_grad=True)
        y = torch.randn(4, 4, requires_grad=True)
        args = (x, y)

        backend = EagerAndRecordGraphs()
        cnt = CompileCounterWithBackend(backend)

        # expected = fn(*args)
        result = torch.compile(fn, backend=cnt)(*args)
~~~

Are we checking something like this raises the same error for each and torch.compile?

File: test/dynamo/test_higher_order_ops.py (hunk 7 above)
Comment (anijain2305) [reply test/dynamo/test_higher_order_ops.py:446]: Added

File: torch/_higher_order_ops/wrap.py (hunk 3 above)
Comment (zou3519) [reply torch/_higher_order_ops/wrap.py:29]: Let's say the function being checkpointed raises an exception. There are two cases:
1. Dynamo graph breaks on exceptions. Then we would expect checkpoint(func, *args) to go the fallback route.
2. Dynamo preserves exceptions in the graph (seems unlikely to me, there's no try-catch syntax in FX). Then we would expect checkpoint(func, *args) to not fallback to eager mode, but still raise an exception later down the line (maybe in AOTAutograd?)

So yes, we would expect an error to be raised for eager and torch.compile. Though I'm more curious about which of the two above cases we're in.

File: torch/_dynamo/variables/torch.py [hunk 11]
```diff
@@ -1182,14 +1185,16 @@ def speculate_branch(branch):
         else:
             unimplemented(f"HigherOrderOperator {self.value.__name__}")
 
+        _, p_kwargs = proxy_args_kwargs([], kwargs)
```
Comment (zou3519) [inline torch/_dynamo/variables/torch.py:1188]: Btw, the kwargs didn't get passed to speculate_subgraph. Is that... okay?

Comment (zou3519) [reply torch/_dynamo/variables/torch.py:1188]: Nvm, these are kwargs to checkpoint.

File: torch/_higher_order_ops/wrap.py [hunk 12]
```diff
@@ -9,5 +10,23 @@ def __call__(self, func, *args):
         result = func(*args)
         return result
 
-
 wrap = Wrap()
+
+class WrapActivationCheckpoint(HigherOrderOperator):
```
Comment (ezyang) [inline torch/_higher_order_ops/wrap.py:15]: This seems like a weird file to put this in, originally, this file was "we can take an arbitrary function and wrap it into a higher order op, for testing". The new class doesn't "wrap" in the conventional sense, it always calls checkpoint. Should this not have gotten its own file in `torch/_higher_order_ops`?
